cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java

```diff
@@ -83,7 +90,66 @@ public AggregatedMetricsEmitter load(EmitterKey key) throws Exception {
   protected final void runOneIteration() throws Exception {
     final long timestamp = TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS);
     LOG.trace("Start log collection for timestamp {}", timestamp);
-    Iterator<MetricValue> metricsItor = getMetrics(timestamp);
+
+    final Iterator<MetricValue> rawMetricsItor = getMetrics(timestamp);
+    Iterator<MetricValue> metricsItor;
+    if (!recordMetaMetrics) {
+      metricsItor = rawMetricsItor;
+    } else {
+      // wrap the raw metrics iterator with an iterator that will publish meta metrics
+      metricsItor = new AbstractIterator<MetricValue>() {
+        private Queue<MetricValue> pendingMetrics = new LinkedList<MetricValue>();
+        private boolean queuedAggregateMetrics = false;
+        private int numMetrics = 0;
+
+        @Override
+        protected MetricValue computeNext() {
+          if (rawMetricsItor.hasNext()) {
+
+            if (pendingMetrics.isEmpty()) {
+              // publish raw metric
+              MetricValue raw = rawMetricsItor.next();
+              numMetrics++;
+
+              // sample processing delay - assumes we mostly receive metrics in time order
+              if (raw.getTimestamp() - timestampLastSampled >= SAMPLING_DELAY_SEC) {
+                timestampLastSampled = raw.getTimestamp();
+
+                long timestampMs = TimeUnit.SECONDS.toMillis(raw.getTimestamp());
+                long nowMs = System.currentTimeMillis();
+                long delayMs = nowMs - timestampMs;
+                MetricValue processDelayMetric = new MetricValue(
+                  ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, "system"),
+                  "metrics.global.processed.delay.ms", timestamp, delayMs, MetricType.GAUGE);
+                pendingMetrics.add(processDelayMetric);
+              }
+
+              return raw;
+            } else {
+              // publish pending metric
+              return pendingMetrics.poll();
+            }
+          } else {
+            if (!queuedAggregateMetrics) {
```

Comment:
I think `queuedAggregateMetrics` is meaningful: If true, that means we queued up the metrics representing the aggregate of the current metrics batch, and thus it's time to send out the queued up aggregate metrics.